mvc/src/main/java/com/interface21/webmvc/servlet/mvc/tobe/AnnotationHandlerMapping.java

```diff
@@ -21,9 +29,64 @@ public AnnotationHandlerMapping(final Object... basePackage) {
 
     public void initialize() {
         log.info("Initialized AnnotationHandlerMapping!");
+        Arrays.stream(basePackage).forEach(base -> {
+            Reflections reflections = new Reflections(base);
+            Set<Class<?>> controllers = reflections.getTypesAnnotatedWith(Controller.class);
+
+            controllers.forEach(this::reflect);
+        });
+    }
+
+    private void reflect(Class<?> controller) {
+        try {
+            Object instance = controller.getConstructor().newInstance();
+            Method[] methods = controller.getDeclaredMethods();
+
+            Arrays.stream(methods).forEach(method -> addHandlerExecution(instance, method));
+
+        } catch (NoSuchMethodException | InvocationTargetException | InstantiationException |
+                 IllegalAccessException e) {
+            throw new RuntimeException("controller 리플렉션 중 실패");
+        }
+    }
+
+    private void addHandlerExecution(Object handlerInstance, Method handlerMethod) {
+
+        String uri = (String) executeMethod("value", handlerMethod);
+        RequestMethod[] requestMethods = (RequestMethod[]) executeMethod("method", handlerMethod);
+
+        if (requestMethods.length == 0) {
+            requestMethods = RequestMethod.values();
+        }
+        log.info("uri = {}, method = {} 등록!", uri, requestMethods);
+
+        HandlerExecution handlerExecution = new HandlerExecution(handlerInstance, handlerMethod);
+
+        Arrays.stream(requestMethods).forEach(requestMethod -> {
+            HandlerKey handlerKey = new HandlerKey(uri, requestMethod);
+            handlerExecutions.put(handlerKey, handlerExecution);
+        });
+    }
+
+    private Object executeMethod(String methodName, Method method) {
+        try {
+            Annotation annotation = method.getAnnotation(RequestMapping.class);
+            return annotation.getClass()
+                    .getDeclaredMethod(methodName)
+                    .invoke(annotation);
+        } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
+            throw new RuntimeException("controller method 리플렉션 중 실패");
+        }
     }
 
     public Object getHandler(final HttpServletRequest request) {
-        return null;
+        String uri = request.getRequestURI();
+        RequestMethod requestMethod = RequestMethod.valueOf(request.getMethod());
+        HandlerKey handlerKey = new HandlerKey(uri, requestMethod);
+
+        if (handlerExecutions.containsKey(handlerKey)) {
+            return handlerExecutions.get(handlerKey);
+        }
+        throw new UnsupportedOperationException("지원하지 않는 요청입니다.");
```

Comment:
```suggestion
        if (!handlerExecutions.containsKey(handlerKey)) {
            throw new UnsupportedOperationException("지원하지 않는 요청입니다.");
        }
        return handlerExecutions.get(handlerKey);
```
(제안) 사소한 요청사항이긴 한데 위와 같이 순서가 바뀌었으면 좋겠어요!! getHandler의 기본값이 예외처리인 것처럼 보여서요😅

Reply:
그런 관점으로 해석할 수도 있겠네요🧐 
저는 가능한 부정 조건문을 사용하지 않도록 코드를 작성하려고 해요. 조건에 부정을 걸면 직관적이지가 않거든요.
절대적인 기준은 아니지만, 참고해보셔도 좋을 것 같아요! 제가 작성한 글은 아닙니다ㅎㅎ [[TIP] 부정조건문은 의미해석이 쉽지 않다.](https://kghworks.tistory.com/66)

부정문을 긍정문으로 바꾸려고 메서드로 한번 더 감싸는건 오히려 코드 복잡도가 늘어나서 그대로 두었습니다~!

Reply:
오 이런 관점도 있었군요!!
확실히 그 얘기를 듣고 보니 제가 제안했던 코드가 더 복잡해보이는 것 같기도 하네요😂